### PR TITLE
[shoya] Sparse top-k slice attention — cross-dataset

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver architecture uses all-pairs attention across physics-informed slice tokens. For large meshes (DrivAerML: 50 000 surface points → many slices), most slice-to-slice attention weights are near-zero — only spatially proximate or physically coupled slices meaningfully interact. By keeping only the top-k attention connections per query (sparse attention), we reduce the quadratic O(S²) attention cost to O(S·k), allowing either more slices per forward pass or more layers within the same compute budget. Better slice coverage → better spatial resolution of flow features → lower surface error.

## Prior Art / Why Now

Sparse attention (Longformer, BigBird, Routing Transformers) has been extensively studied for sequences, but has not been tested in this programme on slice tokens. The key insight here is that physics-informed slices cluster geometrically — the leading edge slices are not strongly coupled to wake slices — making top-k sparsity a physically motivated approximation, not just a compute trick.

## Implementation

If `--topk-attention K` flag exists, use it. Otherwise implement as follows:

In the Transolver slice attention module, after computing attention logits `scores = Q @ K.T / sqrt(d_k)` (shape `[B, H, S, S]`), apply top-k masking before softmax:

```python
def topk_sparse_attention(q, k, v, topk=16):
    """Replace full softmax attention with top-k sparse attention."""
    scale = q.shape[-1] ** -0.5
    scores = torch.matmul(q, k.transpose(-2, -1)) * scale  # [B, H, S, S]
    
    # Keep only top-k logits per query, mask the rest to -inf
    if topk < scores.shape[-1]:
        topk_vals, topk_idx = torch.topk(scores, k=topk, dim=-1)
        mask = torch.full_like(scores, float('-inf'))
        mask.scatter_(-1, topk_idx, topk_vals)
        scores = mask
    
    attn = torch.softmax(scores, dim=-1)
    return torch.matmul(attn, v)
```

## Baselines to Beat

| Dataset | Metric | Current Best | Best PR |
|---|---|---|---|
| TandemFoil | val_primary/surface_pressure_mae | **26.06** | #2887 |
| TandemFoil Paper | val_primary/surface_pressure_mae | not yet established | — |
| AirfRANS | val_primary/surface_pressure_mae | **0.000598** | #2906 |
| DrivAerML | val_primary/surface_drag_coefficient_error | **3.997%** | #2898 |

## Ablation Plan

Test **two values of k** per dataset to find the sweet spot. Use the same base config as current best for each dataset:

- **k=16**: aggressive sparsity — only 16 slices per query attend to each other
- **k=32**: moderate sparsity — 50% density (baseline uses 64 slices = full attention = k=64)

### TandemFoil

```bash
# k=16 (aggressive)
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --topk-attention 16 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset

# k=32 (moderate)
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --topk-attention 32 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset
```

### TandemFoil Paper

```bash
# k=16
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --topk-attention 16 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset

# k=32
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --topk-attention 32 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset
```

### AirfRANS

```bash
# k=16
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw --lr 5e-4 --cosine-t-max 50 --weight-decay 1e-4 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 4 \
  --enable-fourier \
  --topk-attention 16 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset

# k=32
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --optimizer adamw --lr 5e-4 --cosine-t-max 50 --weight-decay 1e-4 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 4 \
  --enable-fourier \
  --topk-attention 32 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset
```

### DrivAerML

```bash
# k=32 (DrivAerML has 50k surface points → more slices, need more connectivity)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --topk-attention 32 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset

# k=64 (moderate — DrivAerML baseline uses 8H so full-attention would be S×S, k=64 is aggressive cut)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --topk-attention 64 \
  --epochs 999 \
  --wandb_group sparse-topk-cross-dataset
```

## If --topk-attention flag doesn't exist

Implement the `topk_sparse_attention` function above and integrate it into the Transolver attention module. Locate the slice attention computation (likely a method called `attention`, `slice_attention`, or `physics_attention` in the model code). Replace the softmax call with the top-k masked version. Make k a command-line argument via `argparse`:

```python
parser.add_argument('--topk-attention', type=int, default=None,
    help='If set, restrict slice attention to top-k keys per query (sparse attention)')
```

Pass `topk=args.topk_attention` through to the model constructor so it can use the value in the attention module. If `--topk-attention` is not set (None), fall back to standard full attention — this ensures backward compatibility.

## What to Report

In your results comment, please include:
1. Best primary validation metric per dataset for each k value tested
2. Whether sparsity improved or degraded performance vs full attention (k=64)
3. W&B run IDs for all runs
4. If possible: memory usage comparison (sparse vs dense) — useful for future experiments combining sparsity with larger models

## Suggested Follow-ups

If top-k sparsity helps on DrivAerML:
- Try using sparsity to enable more slices (e.g., increase `--model-slices` from 64 to 128 or 256 while keeping k=32)
- Combine with MQA/GQA (PRs #2988/#2989) — shared KV heads + sparse attention is doubly memory-efficient
- Try learned sparsity patterns (e.g., routing top-k based on geometric proximity of slice centroids)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### Summary Table (experiments still running — snapshot at ~100 min)

| Dataset | k | Slices | k/S | Epochs | Best val_primary | Baseline | Gap | W&B Run |
|---------|---|--------|-----|--------|-----------------|----------|-----|---------|
| TandemFoil | 16 | 64 | 25% | 25 | 49.61 (ep24) | 26.06 | 1.9× | `pkspydyn` |
| TandemFoil | 32 | 64 | 50% | 22 | 50.25 (ep21) | 26.06 | 1.9× | `1hd3pdpe` |
| AirfRANS | 16 | 64 | 25% | 103 | 0.0084 (ep103) | 0.000598 | 14× | `zgzwy5lo` |
| AirfRANS | 32 | 64 | 50% | 102 | 0.0108 (ep90) | 0.000598 | 18× | `tfmdw3zg` |
| DrivAerML | 32 | 96 | 33% | 63 | 8.82% (ep63) | 3.997% | 2.2× | `evzmuij0` |
| DrivAerML | 64 | 96 | 67% | 63 | 9.88% (ep59) | 3.997% | 2.5× | `qtpffy4s` |

**Primary metrics**: TF=`surface_pressure_mae`, AF=`surface_mse`, DR=`surface_rel_l2_pct`

### Commands

```bash
# TandemFoil k=16 (GPU 0)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --topk-attention 16 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/tandemfoil-topk-16 --agent shoya

# TandemFoil k=32 (GPU 1)
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --topk-attention 32 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/tandemfoil-topk-32 --agent shoya

# AirfRANS k=16 (GPU 2)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 50 --weight-decay 1e-4 --ema-decay 0.999 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 4 --enable-fourier --topk-attention 16 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/airfrans-topk-16 --agent shoya

# AirfRANS k=32 (GPU 3)
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 50 --weight-decay 1e-4 --ema-decay 0.999 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 4 --enable-fourier --topk-attention 32 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/airfrans-topk-32 --agent shoya

# DrivAerML k=32 (GPUs 4-5)
CUDA_VISIBLE_DEVICES=4,5 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --topk-attention 32 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/drivaerml-topk-32 --agent shoya

# DrivAerML k=64 (GPUs 6-7)
CUDA_VISIBLE_DEVICES=6,7 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --topk-attention 64 --epochs 999 --wandb-group sparse-topk-cross-dataset --wandb-name shoya/drivaerml-topk-64 --agent shoya
```

### Peak Memory Usage

| GPU | Experiment | VRAM |
|-----|-----------|------|
| 0 | TF k=16 | 42.8 GB |
| 1 | TF k=32 | 34.4 GB |
| 2 | AF k=16 | 24.2 GB |
| 3 | AF k=32 | 39.8 GB |
| 4-5 | DR k=32 | 6.1 GB per GPU |
| 6-7 | DR k=64 | 5.9 GB per GPU |

### What Happened

**Sparse top-k attention significantly degrades convergence across all three datasets.** After 63–103 epochs (100+ minutes of training), all experiments remain 1.9–14× worse than their baselines. The convergence curves are still descending but the rate is logarithmic and slowing substantially.

Key observations:

1. **More aggressive sparsity paradoxically outperforms less aggressive on 2/3 datasets:**
   - AirfRANS: k=16 (25% keys) best=0.0084 **beats** k=32 (50% keys) best=0.0108
   - DrivAerML: k=32 (33% keys) best=8.82% **beats** k=64 (67% keys) best=9.88%
   - TandemFoil: k=16 ≈ k=32 (both ~50, too early to differentiate)
   
   This suggests top-k may act as a regularizer, but the overall effect is still strongly negative vs. baseline.

2. **The Transolver's "slice" attention operates over a small number of slices (64–96), not individual tokens.** Top-k on such a small key space is an aggressive constraint — e.g., k=16 on S=64 slices means each query only attends to 25% of physics-grouped slices. This is fundamentally different from top-k on standard token-level attention with thousands of tokens.

3. **AirfRANS shows the largest gap (14–18×)**, likely because the full-field (surface+volume) prediction task benefits most from full slice interaction. DrivAerML (surface-only, 2.2–2.5×) and TandemFoil (surface-only, 1.9×) show smaller gaps.

4. **Note on config differences:** The AF experiments use `lr=5e-4, cosine-t-max=50` (per PR) vs baseline's `lr=6e-4, cosine-t-max=10`. The DR experiments use `cosine-t-max=30` vs baseline's config. These schedule differences partially confound the comparison, as baseline uses more cosine restart cycles in the same epoch budget.

### Suggested Follow-ups

1. **Top-k is likely the wrong sparsity target for Transolver.** The slice dimension is already compact (64–96). Sparsity would be better applied to the slice *assignment* weights (the `slice_weights` softmax over tokens → slices) rather than the QK attention between slices.
2. If sparse attention is still desired, test **top-k on a model with many more slices** (e.g., 256+) where the key space is large enough for meaningful sparsification.
3. A controlled experiment comparing topk vs no-topk **with identical hyperparameters** would isolate the effect of sparsity from the schedule differences.

**Comment by tcapelle** [comment]
ADVISOR: Closing as a dead end.

Top-k sparsity on Transolver's slice attention degraded all datasets by 2–18x vs current baselines. The slice space in this model is S=64–96 tokens, which is already small enough that full attention is near-optimal — restricting to top-k=16 or top-k=32 removes information that matters. Sparse attention is a throughput optimization for long sequences (S >> 1000); it is not an inductive bias improvement for this regime.

Results vs current baselines (TF=26.06 #2887, AF=0.000598 #2906, DM=3.997% #2898):
- TandemFoil: 31.2 (top-k=16), 29.4 (top-k=32) — both worse than 26.06
- AirfRANS: 0.00089, 0.00076 — both worse than 0.000598
- DrivAerML: 5.41%, 4.82% — both worse than 3.997%

No path to improvement here. Closing.
